# **Cross-selling Recommendations for Banking Products**

## **Problem Description:**
XYZ Credit Union, a prominent financial institution in Latin America, has been successful in selling various banking products including credit cards, deposit accounts, retirement accounts, and safe deposit boxes. However, despite the initial success, the existing customers do not show any inclination to purchase more than one product. This indicates that the bank is not effectively utilizing the opportunity of cross-selling to its customers. To address this issue, XYZ Credit Union has sought the assistance of ABC Analytics to analyze the data and provide actionable insights to increase cross-selling.

## **Business Understanding:**
Based on the given problem statement, while XYZ Credit Union has been successful in selling a variety of banking products in Latin America, they have not been able to capitalize on cross-selling opportunities. Despite having a large customer base, the bank's existing customers do not purchase more than one product, which is indicative of a lack of effective cross-selling strategies. This project aims to identify the factors that affect cross-selling, understand customer behavior, and propose strategies that can improve the bank's overall performance in selling multiple products to its existing customers.



In [1]:
# Importing required libraries and packages

import pandas as pd
import numpy as np

In [2]:
# Importing the dataset
df = pd.read_csv(r"C:\Users\Oludipe j\Documents\Data Analysis\Data Glacier\Cross Selling\Train.csv", low_memory=False)

df

fecha_dato  ncodpers ind_empleado pais_residencia sexo  age  \
0         2015-01-28   1375586            N              ES    H   35   
1         2015-01-28   1050611            N              ES    V   23   
2         2015-01-28   1050612            N              ES    V   23   
3         2015-01-28   1050613            N              ES    H   22   
4         2015-01-28   1050614            N              ES    V   23   
...              ...       ...          ...             ...  ...  ...   
13647304  2016-05-28   1166765            N              ES    V   22   
13647305  2016-05-28   1166764            N              ES    V   23   
13647306  2016-05-28   1166763            N              ES    H   47   
13647307  2016-05-28   1166789            N              ES    H   22   
13647308  2016-05-28   1550586            N              ES    H   37   

          fecha_alta  ind_nuevo antiguedad  indrel  ... ind_hip_fin_ult1  \
0         2015-01-12        0.0          6     1.0  ...                0   
1         2012-08-10        0.0         35     1.0  ...                0   
2         2012-08-10        0.0         35     1.0  ...                0   
3         2012-08-10        0.0         35     1.0  ...                0   
4         2012-08-10        0.0         35     1.0  ...                0   
...              ...        ...        ...     ...  ...              ...   
13647304  2013-08-14        0.0         33     1.0  ...                0   
13647305  2013-08-14        0.0         33     1.0  ...                0   
13647306  2013-08-14        0.0         33     1.0  ...                0   
13647307  2013-08-14        0.0         33     1.0  ...                0   
13647308  2016-05-13        1.0          0     1.0  ...                0   

         ind_plan_fin_ult1 ind_pres_fin_ult1 ind_reca_fin_ult1  \
0                        0                 0                 0   
1                        0                 0                 0   
2                        0                 0                 0   
3                        0                 0                 0   
4                        0                 0                 0   
...                    ...               ...               ...   
13647304                 0                 0                 0   
13647305                 0                 0                 0   
13647306                 0                 0                 0   
13647307                 0                 0                 0   
13647308                 0                 0                 0   

         ind_tjcr_fin_ult1 ind_valo_fin_ult1 ind_viv_fin_ult1 ind_nomina_ult1  \
0                        0                 0                0             0.0   
1                        0                 0                0             0.0   
2                        0                 0                0             0.0   
3                        0                 0                0             0.0   
4                        0                 0                0             0.0   
...                    ...               ...              ...             ...   
13647304                 0                 0                0             0.0   
13647305                 0                 0                0             0.0   
13647306                 0                 0                0             0.0   
13647307                 0                 0                0             0.0   
13647308                 0                 0                0             0.0   

          ind_nom_pens_ult1  ind_recibo_ult1  
0                       0.0                0  
1                       0.0                0  
2                       0.0                0  
3                       0.0                0  
4                       0.0                0  
...                     ...              ...  
13647304                0.0                0  
13647305                0.0                0  
13647306                0.0    

In [3]:
# Set to display all columns
pd.set_option('display.max_columns', None)

df.head(3)

fecha_dato  ncodpers ind_empleado pais_residencia sexo  age  \
0         2015-01-28   1375586            N              ES    H   35   
1         2015-01-28   1050611            N              ES    V   23   
2         2015-01-28   1050612            N              ES    V   23   
3         2015-01-28   1050613            N              ES    H   22   
4         2015-01-28   1050614            N              ES    V   23   
...              ...       ...          ...             ...  ...  ...   
13647304  2016-05-28   1166765            N              ES    V   22   
13647305  2016-05-28   1166764            N              ES    V   23   
13647306  2016-05-28   1166763            N              ES    H   47   
13647307  2016-05-28   1166789            N              ES    H   22   
13647308  2016-05-28   1550586            N              ES    H   37   

          fecha_alta  ind_nuevo antiguedad  indrel ult_fec_cli_1t indrel_1mes  \
0         2015-01-12        0.0          6     1.0            NaN         1.0   
1         2012-08-10        0.0         35     1.0            NaN           1   
2         2012-08-10        0.0         35     1.0            NaN           1   
3         2012-08-10        0.0         35     1.0            NaN           1   
4         2012-08-10        0.0         35     1.0            NaN           1   
...              ...        ...        ...     ...            ...         ...   
13647304  2013-08-14        0.0         33     1.0            NaN           1   
13647305  2013-08-14        0.0         33     1.0            NaN           1   
13647306  2013-08-14        0.0         33     1.0            NaN           1   
13647307  2013-08-14        0.0         33     1.0            NaN         1.0   
13647308  2016-05-13        1.0          0     1.0            NaN         NaN   

         tiprel_1mes indresi indext conyuemp canal_entrada indfall  tipodom  \
0                  A       S      N      NaN           KHL       N      1.0   
1                  I       S      S      NaN           KHE       N      1.0   
2                  I       S      N      NaN           KHE       N      1.0   
3                  I       S      N      NaN           KHD       N      1.0   
4                  A       S      N      NaN           KHE       N      1.0   
...              ...     ...    ...      ...           ...     ...      ...   
13647304           I       S      N      NaN           KHE       N      1.0   
13647305           I       S      N      NaN           KHE       N      1.0   
13647306           A       S      N      NaN           KHE       N      1.0   
13647307           I       S      N      NaN           KHE       N      1.0   
13647308         NaN       S      N      NaN           NaN       N      1.0   

          cod_prov      nomprov  ind_actividad_cliente      renta  \
0             29.0       MALAGA                    1.0   87218.10   
1             13.0  CIUDAD REAL                    0.0   35548.74   
2             13.0  CIUDAD REAL                    0.0  122179.11   
3             50.0     ZARAGOZA                    0.0  119775.54   
4             50.0     ZARAGOZA                    1.0        NaN   
...            ...          ...                    ...        ...   
13647304      50.0     ZARAGOZA                    0.0   43912.17   
13647305      26.0    RIOJA, LA                    0.0   23334.99   
13647306      50.0     ZARAGOZA                    1.0        NaN   
13647307      50.0     ZARAGOZA                    0.0  199592.82   
13647308      28.0       MADRID                    1.0        NaN   

                    segmento  ind_ahor_fin_ult1  ind_aval_fin_ult1  \
0          02 - PARTICULARES                  0                  0   
1         03 - UNIVERSITARIO                  0                  0   
2         03 - UNIVERSITARIO                  0                  0   
3         03 - UNIVERSITARIO                  0                  0   
4         03 - UNIVERSITARIO    

## **Data Understanding:**
The dataset that has been provided includes comprehensive customer information such as age, gender, and country of residence, alongside the various bank products that they currently own, such as credit cards, deposit accounts, retirement accounts, safe deposit boxes, and more. In total, the dataset comprises 48 features (columns) and 13647308 observations (rows), offering a large volume of data for analysis and providing valuable insights into the bank's customer base and product preferences.

## **Data Cleaning:**
As a data analyst, I recognize the importance of data cleaning to ensure that the data is accurate and reliable for any downstream analysis or use. XYZ Credit Union is also a Latina American American bank which made a lot of the column names and data values to be written in spanish, these words or abbrieviations had to be translated to English for better understanding.


In [4]:
# Checking all column names

print(df.columns)

Index(['fecha_dato', 'ncodpers', 'ind_empleado', 'pais_residencia', 'sexo',
       'age', 'fecha_alta', 'ind_nuevo', 'antiguedad', 'indrel',
       'ult_fec_cli_1t', 'indrel_1mes', 'tiprel_1mes', 'indresi', 'indext',
       'conyuemp', 'canal_entrada', 'indfall', 'tipodom', 'cod_prov',
       'nomprov', 'ind_actividad_cliente', 'renta', 'segmento',
       'ind_ahor_fin_ult1', 'ind_aval_fin_ult1', 'ind_cco_fin_ult1',
       'ind_cder_fin_ult1', 'ind_cno_fin_ult1', 'ind_ctju_fin_ult1',
       'ind_ctma_fin_ult1', 'ind_ctop_fin_ult1', 'ind_ctpp_fin_ult1',
       'ind_deco_fin_ult1', 'ind_deme_fin_ult1', 'ind_dela_fin_ult1',
       'ind_ecue_fin_ult1', 'ind_fond_fin_ult1', 'ind_hip_fin_ult1',
       'ind_plan_fin_ult1', 'ind_pres_fin_ult1', 'ind_reca_fin_ult1',
       'ind_tjcr_fin_ult1', 'ind_valo_fin_ult1', 'ind_viv_fin_ult1',
       'ind_nomina_ult1', 'ind_nom_pens_ult1', 'ind_recibo_ult1'],
      dtype='object')


In [5]:
# Changing all column names to appropriate and descriptive ones

df = df.rename(columns={'fecha_dato': 'data_date', 'ncodpers': 'customer_id', 'ind_empleado': 'employee_index', 'pais_residencia': 'country_of_residence', 'sexo': 'gender',
       'fecha_alta': 'date_joined', 'ind_nuevo': 'new_customer', 'antiguedad': 'customer_seniority', 'indrel': 'primary_index',
       'ult_fec_cli_1t': 'last_date_as_primary', 'indrel_1mes': 'customer_type', 'tiprel_1mes': 'customer_relation', 'indresi': 'resident', 'indext': 'foreigner',
       'conyuemp': 'employee_spouse', 'canal_entrada': 'channel', 'indfall': 'deceased', 'tipodom': 'address_type', 'cod_prov': 'province_code',
       'nomprov': 'province_name', 'ind_actividad_cliente': 'activity_index', 'renta': 'gross_household_income', 'segmento': 'customer_segment',
       'ind_ahor_fin_ult1': 'saving_acc', 'ind_aval_fin_ult1': 'guarantees', 'ind_cco_fin_ult1': 'current_acc',
       'ind_cder_fin_ult1': 'derivative_acc', 'ind_cno_fin_ult1': 'payroll_acc', 'ind_ctju_fin_ult1': 'junior_acc',
       'ind_ctma_fin_ult1': 'more_particular_acc', 'ind_ctop_fin_ult1': 'particular_acc', 'ind_ctpp_fin_ult1': 'particular_plus_acc',
       'ind_deco_fin_ult1': 'short_term_deposits', 'ind_deme_fin_ult1': 'medium_term_deposits', 'ind_dela_fin_ult1': 'long_term_deposits',
       'ind_ecue_fin_ult1': 'e_account', 'ind_fond_fin_ult1': 'funds', 'ind_hip_fin_ult1': 'mortgage',
       'ind_plan_fin_ult1': 'pensions_plan', 'ind_pres_fin_ult1': 'loans', 'ind_reca_fin_ult1': 'taxes',
       'ind_tjcr_fin_ult1': 'credit_card', 'ind_valo_fin_ult1': 'securities', 'ind_viv_fin_ult1': 'home_acc',
       'ind_nomina_ult1': 'payroll', 'ind_nom_pens_ult1': 'pensions', 'ind_recibo_ult1': 'direct_debit'})

print(df.columns)

Index(['data_date', 'customer_id', 'employee_index', 'country_of_residence',
       'gender', 'age', 'date_joined', 'new_customer', 'customer_seniority',
       'primary_index', 'last_date_as_primary', 'customer_type',
       'customer_relation', 'resident', 'foreigner', 'employee_spouse',
       'channel', 'deceased', 'address_type', 'province_code', 'province_name',
       'activity_index', 'gross_household_income', 'customer_segment',
       'saving_acc', 'guarantees', 'current_acc', 'derivative_acc',
       'payroll_acc', 'junior_acc', 'more_particular_acc', 'particular_acc',
       'particular_plus_acc', 'short_term_deposits', 'medium_term_deposits',
       'long_term_deposits', 'e_account', 'funds', 'mortgage', 'pensions_plan',
       'loans', 'taxes', 'credit_card', 'securities', 'home_acc', 'payroll',
       'pensions', 'direct_debit'],
      dtype='object')


In [6]:
# Viewing Table
df.head(2)

data_date  customer_id employee_index country_of_residence gender  age  \
0  2015-01-28      1375586              N                   ES      H   35   
1  2015-01-28      1050611              N                   ES      V   23   
2  2015-01-28      1050612              N                   ES      V   23   

  date_joined  new_customer customer_seniority  primary_index  \
0  2015-01-12           0.0                  6            1.0   
1  2012-08-10           0.0                 35            1.0   
2  2012-08-10           0.0                 35            1.0   

  last_date_as_primary customer_type customer_relation resident foreigner  \
0                  NaN           1.0                 A        S         N   
1                  NaN             1                 I        S         S   
2                  NaN             1                 I        S         N   

  employee_spouse channel deceased  address_type  province_code province_name  \
0             NaN     KHL        N           1.0           29.0        MALAGA   
1             NaN     KHE        N           1.0           13.0   CIUDAD REAL   
2             NaN     KHE        N           1.0           13.0   CIUDAD REAL   

   activity_index  gross_household_income    customer_segment  saving_acc  \
0             1.0                87218.10   02 - PARTICULARES           0   
1             0.0                35548.74  03 - UNIVERSITARIO           0   
2             0.0               122179.11  03 - UNIVERSITARIO           0   

   guarantees  current_acc  derivative_acc  payroll_acc  junior_acc  \
0           0            1               0            0           0   
1           0            1               0            0           0   
2           0            1               0            0           0   

   more_particular_acc  particular_acc  particular_plus_acc  \
0                    0               0                    0   
1                    0               0                    0   
2                    0               0                    0   

   short_term_deposits  medium_term_deposits  long_term_deposits  e_account  \
0                    0                     0                   0          0   
1                    0                     0                   0          0   
2                    0                     0                   0          0   

   funds  mortgage  pensions_plan  loans  taxes  credit_card  securities  \
0      0         0              0      0      0            0           0   
1      0         0              0      0      0            0           0   
2      0         0              0      0      0            0           0   

   home_acc  payroll  pensions  direct_debit  
0         0      0.0       0.0             0  
1         0      0.0       0.0             0  
2         0      0.0       0.0             0

In [9]:
# Converting values in the gender column from "H" and "V" to "M" and "F" to standardize the data.

df['gender'] = df['gender'].replace({'H': 'M', 'V': 'F'})

df.head(3)

data_date  customer_id employee_index country_of_residence gender  age  \
0  2015-01-28      1375586              N                   ES      M   35   
1  2015-01-28      1050611              N                   ES      F   23   
2  2015-01-28      1050612              N                   ES      F   23   
3  2015-01-28      1050613              N                   ES      M   22   
4  2015-01-28      1050614              N                   ES      F   23   

  date_joined  new_customer customer_seniority  primary_index  \
0  2015-01-12           0.0                  6            1.0   
1  2012-08-10           0.0                 35            1.0   
2  2012-08-10           0.0                 35            1.0   
3  2012-08-10           0.0                 35            1.0   
4  2012-08-10           0.0                 35            1.0   

  last_date_as_primary customer_type customer_relation resident foreigner  \
0                  NaN           1.0                 A      Yes        No   
1                  NaN             1                 I      Yes       Yes   
2                  NaN             1                 I      Yes        No   
3                  NaN             1                 I      Yes        No   
4                  NaN             1                 A      Yes        No   

  employee_spouse channel deceased  address_type  province_code province_name  \
0             NaN     KHL       No           1.0           29.0        MALAGA   
1             NaN     KHE       No           1.0           13.0   CIUDAD REAL   
2             NaN     KHE       No           1.0           13.0   CIUDAD REAL   
3             NaN     KHD       No           1.0           50.0      ZARAGOZA   
4             NaN     KHE       No           1.0           50.0      ZARAGOZA   

   activity_index  gross_household_income    customer_segment  saving_acc  \
0             1.0                87218.10   02 - PARTICULARES           0   
1             0.0                35548.74  03 - UNIVERSITARIO           0   
2             0.0               122179.11  03 - UNIVERSITARIO           0   
3             0.0               119775.54  03 - UNIVERSITARIO           0   
4             1.0                     NaN  03 - UNIVERSITARIO           0   

   guarantees  current_acc  derivative_acc  payroll_acc  junior_acc  \
0           0            1               0            0           0   
1           0            1               0            0           0   
2           0            1               0            0           0   
3           0            0               0            0           0   
4           0            1               0            0           0   

   more_particular_acc  particular_acc  particular_plus_acc  \
0                    0               0                    0   
1                    0               0                    0   
2                    0               0                    0   
3                    0               0                    0   
4                    0               0                    0   

   short_term_deposits  medium_term_deposits  long_term_deposits  e_account  \
0                    0                     0                   0          0   
1                    0                     0                   0          0   
2                    0                     0                   0          0   
3                    1                     0                   0          0   
4                    0                     0                   0          0   

   funds  mortgage  pensions_plan  loans  taxes  credit_card  securities  \
0      0         0              0      0      0            0           0   
1      0         0              0      0      0            0           0   
2      0         0              0      0      0            0           0   
3      0         0              0      0      0            0           0   
4      0         0              0      0      0            0          

In [10]:
# Converting the "S" and "N" values to "Yes" and "No" to make them more explicit.
df['resident'] = df['resident'].replace({'S': 'Yes', 'N': 'No'})

df['foreigner'] = df['foreigner'].replace({'S': 'Yes', 'N': 'No'})

df['deceased'] = df['deceased'].replace({'S': 'Yes', 'N': 'No'})

df.head(3)

data_date  customer_id employee_index country_of_residence gender  age  \
0  2015-01-28      1375586              N                   ES      M   35   
1  2015-01-28      1050611              N                   ES      F   23   
2  2015-01-28      1050612              N                   ES      F   23   

  date_joined  new_customer customer_seniority  primary_index  \
0  2015-01-12           0.0                  6            1.0   
1  2012-08-10           0.0                 35            1.0   
2  2012-08-10           0.0                 35            1.0   

  last_date_as_primary customer_type customer_relation resident foreigner  \
0                  NaN           1.0                 A      Yes        No   
1                  NaN             1                 I      Yes       Yes   
2                  NaN             1                 I      Yes        No   

  employee_spouse channel deceased  address_type  province_code province_name  \
0             NaN     KHL       No           1.0           29.0        MALAGA   
1             NaN     KHE       No           1.0           13.0   CIUDAD REAL   
2             NaN     KHE       No           1.0           13.0   CIUDAD REAL   

   activity_index  gross_household_income    customer_segment  saving_acc  \
0             1.0                87218.10   02 - PARTICULARES           0   
1             0.0                35548.74  03 - UNIVERSITARIO           0   
2             0.0               122179.11  03 - UNIVERSITARIO           0   

   guarantees  current_acc  derivative_acc  payroll_acc  junior_acc  \
0           0            1               0            0           0   
1           0            1               0            0           0   
2           0            1               0            0           0   

   more_particular_acc  particular_acc  particular_plus_acc  \
0                    0               0                    0   
1                    0               0                    0   
2                    0               0                    0   

   short_term_deposits  medium_term_deposits  long_term_deposits  e_account  \
0                    0                     0                   0          0   
1                    0                     0                   0          0   
2                    0                     0                   0          0   

   funds  mortgage  pensions_plan  loans  taxes  credit_card  securities  \
0      0         0              0      0      0            0           0   
1      0         0              0      0      0            0           0   
2      0         0              0      0      0            0           0   

   home_acc  payroll  pensions  direct_debit  
0         0      0.0       0.0             0  
1         0      0.0       0.0             0  
2         0      0.0       0.0             0

In [11]:
# Converting the values in a customer_segment column from '01 - TOP' to VIP, '02 - PARTICULARES' to Individuals,
#and 03 - UNIVERSITARIO to College Graduate.

# Define a dictionary to map the old values to the new values
mapping_dict = {
    '01 - TOP': 'VIP',
    '02 - PARTICULARES': 'Individuals',
    '03 - UNIVERSITARIO': 'College Graduate'
}

# Replace the old values in the 'customer_segment' column with the new values using the mapping dictionary
df['customer_segment'] = df['customer_segment'].replace(mapping_dict)

df.head(3)

data_date  customer_id employee_index country_of_residence gender  age  \
0  2015-01-28      1375586              N                   ES      M   35   
1  2015-01-28      1050611              N                   ES      F   23   
2  2015-01-28      1050612              N                   ES      F   23   

  date_joined  new_customer customer_seniority  primary_index  \
0  2015-01-12           0.0                  6            1.0   
1  2012-08-10           0.0                 35            1.0   
2  2012-08-10           0.0                 35            1.0   

  last_date_as_primary customer_type customer_relation resident foreigner  \
0                  NaN           1.0                 A      Yes        No   
1                  NaN             1                 I      Yes       Yes   
2                  NaN             1                 I      Yes        No   

  employee_spouse channel deceased  address_type  province_code province_name  \
0             NaN     KHL       No           1.0           29.0        MALAGA   
1             NaN     KHE       No           1.0           13.0   CIUDAD REAL   
2             NaN     KHE       No           1.0           13.0   CIUDAD REAL   

   activity_index  gross_household_income  customer_segment  saving_acc  \
0             1.0                87218.10       Individuals           0   
1             0.0                35548.74  College Graduate           0   
2             0.0               122179.11  College Graduate           0   

   guarantees  current_acc  derivative_acc  payroll_acc  junior_acc  \
0           0            1               0            0           0   
1           0            1               0            0           0   
2           0            1               0            0           0   

   more_particular_acc  particular_acc  particular_plus_acc  \
0                    0               0                    0   
1                    0               0                    0   
2                    0               0                    0   

   short_term_deposits  medium_term_deposits  long_term_deposits  e_account  \
0                    0                     0                   0          0   
1                    0                     0                   0          0   
2                    0                     0                   0          0   

   funds  mortgage  pensions_plan  loans  taxes  credit_card  securities  \
0      0         0              0      0      0            0           0   
1      0         0              0      0      0            0           0   
2      0         0              0      0      0            0           0   

   home_acc  payroll  pensions  direct_debit  
0         0      0.0       0.0             0  
1         0      0.0       0.0             0  
2         0      0.0       0.0             0

In [28]:
# Check data types for all columns
print(df.dtypes)

data_date                         object
customer_id                        int64
employee_index                    object
country_of_residence              object
gender                            object
age                               object
date_joined                       object
new_customer                     float64
customer_seniority                object
primary_index                    float64
last_date_as_primary              object
customer_type                     object
customer_relation                 object
resident                          object
foreigner                         object
employee_spouse                   object
channel                           object
deceased                          object
address_type                     float64
province_code                    float64
province_name                     object
activity_index                   float64
gross_household_income           float64
customer_segment                  object
saving_acc      

In [31]:
# Replace invalid or missing values with NaN
df['age'] = df['age'].replace(' NA', np.nan)

# Convert the 'age' column from object to float
df['age'] = df['age'].astype(float)

# Check data types for all columns
print(df.dtypes)

data_date                         object
customer_id                        int64
employee_index                    object
country_of_residence              object
gender                            object
age                              float64
date_joined                       object
new_customer                     float64
customer_seniority                object
primary_index                    float64
last_date_as_primary              object
customer_type                     object
customer_relation                 object
resident                          object
foreigner                         object
employee_spouse                   object
channel                           object
deceased                          object
address_type                     float64
province_code                    float64
province_name                     object
activity_index                   float64
gross_household_income           float64
customer_segment                  object
saving_acc      

In [12]:
#Check for missing data

print(df.isnull().sum())

data_date                        0
customer_id                      0
employee_index               27734
country_of_residence         27734
gender                       27804
age                              0
date_joined                  27734
new_customer                 27734
customer_seniority               0
primary_index                27734
last_date_as_primary      13622516
customer_type               149781
customer_relation           149781
resident                     27734
foreigner                    27734
employee_spouse           13645501
channel                     186126
deceased                     27734
address_type                 27735
province_code                93591
province_name                93591
activity_index               27734
gross_household_income     2794375
customer_segment            189368
saving_acc                       0
guarantees                       0
current_acc                      0
derivative_acc                   0
payroll_acc         

In [13]:
#Checking for percentage missing data

for col in df.columns:
    pct_missing = np.mean(df[col].isnull())
    print('{} - {}%'.format(col, pct_missing))

data_date - 0.0%
customer_id - 0.0%
employee_index - 0.002032195504622926%
country_of_residence - 0.002032195504622926%
gender - 0.002037324720939491%
age - 0.0%
date_joined - 0.002032195504622926%
new_customer - 0.002032195504622926%
customer_seniority - 0.0%
primary_index - 0.002032195504622926%
last_date_as_primary - 0.9981833048551916%
customer_type - 0.010975130701591061%
customer_relation - 0.010975130701591061%
resident - 0.002032195504622926%
foreigner - 0.002032195504622926%
employee_spouse - 0.999867519669995%
channel - 0.01363829308767025%
deceased - 0.002032195504622926%
address_type - 0.002032268779141734%
province_code - 0.006857835489765785%
province_name - 0.006857835489765785%
activity_index - 0.002032195504622926%
gross_household_income - 0.20475648349429182%
customer_segment - 0.013875849077646003%
saving_acc - 0.0%
guarantees - 0.0%
current_acc - 0.0%
derivative_acc - 0.0%
payroll_acc - 0.0%
junior_acc - 0.0%
more_particular_acc - 0.0%
particular_acc - 0.0%
particul

In [32]:
# Drop rows with missing data in the gender column
df = df.dropna(subset=['gender'])


In [33]:
#Check for missing data

print(df.isnull().sum())

data_date                        0
customer_id                      0
employee_index                   0
country_of_residence             0
gender                           0
age                              0
date_joined                      0
new_customer                     0
customer_seniority               0
primary_index                    0
last_date_as_primary      13594712
customer_type               122047
customer_relation           122047
resident                         0
foreigner                        0
employee_spouse           13617697
channel                     158391
deceased                         0
address_type                     1
province_code                65857
province_name                65857
activity_index                   0
gross_household_income     2766607
customer_segment            161633
saving_acc                       0
guarantees                       0
current_acc                      0
derivative_acc                   0
payroll_acc         

In [82]:
#Converting the 'data_date' column to a datetime format and assign to a new 'data_datetime' column

df['data_datetime'] = pd.to_datetime(df['data_date'])

df

C:\Users\Oludipe j\AppData\Local\Temp\ipykernel_11756\2062779106.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['data_datetime'] = pd.to_datetime(df['data_date'])


data_date  customer_id employee_index country_of_residence gender  \
0        2015-01-28      1375586              N                   ES      M   
1        2015-01-28      1050611              N                   ES      F   
2        2015-01-28      1050612              N                   ES      F   
3        2015-01-28      1050613              N                   ES      M   
4        2015-01-28      1050614              N                   ES      F   
...             ...          ...            ...                  ...    ...   
13647304 2016-05-28      1166765              N                   ES      F   
13647305 2016-05-28      1166764              N                   ES      F   
13647306 2016-05-28      1166763              N                   ES      M   
13647307 2016-05-28      1166789              N                   ES      M   
13647308 2016-05-28      1550586              N                   ES      M   

           age date_joined  new_customer customer_seniority  primary_index  \
0         35.0  2015-01-12           0.0                  6            1.0   
1         23.0  2012-08-10           0.0                 35            1.0   
2         23.0  2012-08-10           0.0                 35            1.0   
3         22.0  2012-08-10           0.0                 35            1.0   
4         23.0  2012-08-10           0.0                 35            1.0   
...        ...         ...           ...                ...            ...   
13647304  22.0  2013-08-14           0.0                 33            1.0   
13647305  23.0  2013-08-14           0.0                 33            1.0   
13647306  47.0  2013-08-14           0.0                 33            1.0   
13647307  22.0  2013-08-14           0.0                 33            1.0   
13647308  37.0  2016-05-13           1.0                  0            1.0   

         last_date_as_primary customer_type customer_relation resident  \
0                         NaN           1.0                 A      Yes   
1                         NaN             1                 I      Yes   
2                         NaN             1                 I      Yes   
3                         NaN             1                 I      Yes   
4                         NaN             1                 A      Yes   
...                       ...           ...               ...      ...   
13647304                  NaN             1                 I      Yes   
13647305                  NaN             1                 I      Yes   
13647306                  NaN             1                 A      Yes   
13647307                  NaN           1.0                 I      Yes   
13647308                  NaN           NaN               NaN      Yes   

         foreigner employee_spouse channel deceased  address_type  \
0               No             NaN     KHL       No           1.0   
1              Yes             NaN     KHE       No           1.0   
2               No             NaN     KHE       No           1.0   
3               No             NaN     KHD       No           1.0   
4               No             NaN     KHE       No           1.0   
...            ...             ...     ...      ...           ...   
13647304        No             NaN     KHE       No           1.0   
13647305        No             NaN     KHE       No           1.0   
13647306        No             NaN     KHE       No           1.0   
13647307        No             NaN     KHE       No           1.0   
13647308        No             NaN     NaN       No           1.0   

          province_code province_name  activity_index  gross_household_income  \
0                  29.0        MALAGA             1.0                87218.10   
1                  13.0   CIUDAD REAL             0.0                35548.74   
2                  13.0   CIUDAD REAL             0.0               122179.11   
3                  50.0      ZARAGOZA             0.0               119775.54   
4

In [34]:
# Creating a dataframe with unique customer ID with most recent data date.
df_sorted = df.loc[df.groupby('customer_id')['data_datetime'].idxmax()]

df_sorted

data_date  customer_id employee_index country_of_residence gender  \
13026343  2016-05-28        15889              F                   ES      F   
13026342  2016-05-28        15890              A                   ES      F   
5319232   2015-08-28        15891              N                   ES      M   
13026341  2016-05-28        15892              F                   ES      M   
13026340  2016-05-28        15893              N                   ES      F   
...              ...          ...            ...                  ...    ...   
13336818  2016-05-28      1553685              N                   ES      F   
13336817  2016-05-28      1553686              N                   ES      M   
13336816  2016-05-28      1553687              N                   ES      F   
13336815  2016-05-28      1553688              N                   ES      M   
13336814  2016-05-28      1553689              N                   ES      F   

           age date_joined  new_customer customer_seniority  primary_index  \
13026343  56.0  1995-01-16           0.0                255            1.0   
13026342  63.0  1995-01-16           0.0                256            1.0   
5319232   59.0  2015-07-28           0.0                246           99.0   
13026341  62.0  1995-01-16           0.0                256            1.0   
13026340  63.0  1997-10-03           0.0                256            1.0   
...        ...         ...           ...                ...            ...   
13336818  52.0  2016-05-31           1.0                  0            1.0   
13336817  30.0  2016-05-31           1.0                  0            1.0   
13336816  21.0  2016-05-31           1.0                  0            1.0   
13336815  43.0  2016-05-31           1.0                  0            1.0   
13336814  40.0  2016-05-31           1.0                  0            1.0   

         last_date_as_primary customer_type customer_relation resident  \
13026343                  NaN             1                 A      Yes   
13026342                  NaN             1                 A      Yes   
5319232            2015-08-05             1                 A      Yes   
13026341                  NaN             1                 A      Yes   
13026340                  NaN             1                 A      Yes   
...                       ...           ...               ...      ...   
13336818                  NaN           NaN               NaN      Yes   
13336817                  NaN           NaN               NaN      Yes   
13336816                  NaN           NaN               NaN      Yes   
13336815                  NaN           NaN               NaN      Yes   
13336814                  NaN           NaN               NaN      Yes   

         foreigner employee_spouse channel deceased  address_type  \
13026343        No               N     KAT       No           1.0   
13026342        No               N     KAT       No           1.0   
5319232         No               N     KAT       No           1.0   
13026341        No               N     KAT       No           1.0   
13026340        No               N     KAT       No           1.0   
...            ...             ...     ...      ...           ...   
13336818        No             NaN     NaN       No           1.0   
13336817       Yes             NaN     NaN       No           1.0   
13336816        No             NaN     NaN       No           1.0   
13336815        No             NaN     NaN       No           1.0   
13336814        No             NaN     NaN       No           1.0   

          province_code province_name  activity_index  gross_household_income  \
13026343           28.0        MADRID             1.0               326124.90   
13026342           28.0        MADRID             1.0                71461.20   
5319232            28.0        MADRID             0.0                     NaN   
13026341           28.0        MADRID             1.0               430

This reduced the dataset to 956,645 unique Customer IDs

In [35]:
#Check for missing data in reduced dataset

print(df_sorted.isnull().sum())

data_date                      0
customer_id                    0
employee_index                 0
country_of_residence           0
gender                         0
age                            0
date_joined                    0
new_customer                   0
customer_seniority             0
primary_index                  0
last_date_as_primary      930280
customer_type               7655
customer_relation           7655
resident                       0
foreigner                      0
employee_spouse           949488
channel                    11433
deceased                       0
address_type                   0
province_code               4018
province_name               4018
activity_index                 0
gross_household_income    240201
customer_segment           11685
saving_acc                     0
guarantees                     0
current_acc                    0
derivative_acc                 0
payroll_acc                    0
junior_acc                     0
more_parti

In [36]:
# Checking for percentage missing data in reduced dataset

for col in df_sorted.columns:
    pct_missing = np.mean(df_sorted[col].isnull())
    print('{} - {}%'.format(col, pct_missing))

data_date - 0.0%
customer_id - 0.0%
employee_index - 0.0%
country_of_residence - 0.0%
gender - 0.0%
age - 0.0%
date_joined - 0.0%
new_customer - 0.0%
customer_seniority - 0.0%
primary_index - 0.0%
last_date_as_primary - 0.9796453066472622%
customer_type - 0.008061212562222978%
customer_relation - 0.008061212562222978%
resident - 0.0%
foreigner - 0.0%
employee_spouse - 0.999872579135202%
channel - 0.012039692125917089%
deceased - 0.0%
address_type - 0.0%
province_code - 0.004231215163293524%
province_name - 0.004231215163293524%
activity_index - 0.0%
gross_household_income - 0.2529472656640786%
customer_segment - 0.012305064505496472%
saving_acc - 0.0%
guarantees - 0.0%
current_acc - 0.0%
derivative_acc - 0.0%
payroll_acc - 0.0%
junior_acc - 0.0%
more_particular_acc - 0.0%
particular_acc - 0.0%
particular_plus_acc - 0.0%
short_term_deposits - 0.0%
medium_term_deposits - 0.0%
long_term_deposits - 0.0%
e_account - 0.0%
funds - 0.0%
mortgage - 0.0%
pensions_plan - 0.0%
loans - 0.0%
taxes -

## **Exploratory Data Analysis:**
Exploratory Data Analysis (EDA) is a crucial step in any data analysis project as it helps us to understand the structure of the data, uncover patterns and relationships, and identify potential outliers or anomalies. The goal of EDA is to gain insights into the data that can help inform business decisions.

In this project, we will conduct EDA to identify and analyze key metrics and performance indicators (KPIs) that can provide valuable insights to XYZ Credit Union. We will explore the relationships between customer demographics and their banking behavior, such as the number and types of banking products they own. By analyzing these relationships, we can gain insights into customer preferences and identify potential cross-selling opportunities.

Some of the key metrics and KPIs that we will explore include:

1. Customer demographics: We will examine the age, gender, and country of residence of the bank's customers to understand their distribution across different groups. This will help us to identify potential target markets for cross-selling campaigns.

2. Product ownership: We will explore the distribution of different banking products across the customer base, including credit cards, deposit accounts, retirement accounts, safe deposit boxes, and more. By analyzing the ownership of these products, we can identify potential cross-selling opportunities and tailor our marketing campaigns to specific customer segments.

3. Cross-selling opportunities: We will analyze the co-ownership patterns of different banking products to identify potential cross-selling opportunities. For example, if a large percentage of customers who own credit cards also own deposit accounts, we can infer that there may be an opportunity to cross-sell deposit accounts to credit card customers.

4. Performance metrics: We will calculate various performance metrics such as customer lifetime value, retention rate, and churn rate to understand the bank's overall performance and identify areas for improvement.

Overall, our EDA will provide valuable insights that can help XYZ Credit Union to optimize its cross-selling strategies and improve customer engagement and retention.

In [37]:
#Total number of unique customers
unique_customers = df['customer_id'].nunique()

# Print the result
print(f"Total number of unique customers: {unique_customers}")

Total number of unique customers: 949609


In [38]:
#Total number of customers in the sorted dataframe
customers = df_sorted['customer_id'].nunique()

# Print the result
print(f"Total number of customers: {customers}")

Total number of customers: 949609


In [43]:
# Customers count by gender
gender_counts = df_sorted['gender'].value_counts()

gender_counts = gender_counts.rename({'F': 'Female', 'M': 'Male'})

# Calculate percentage of customers by gender
total_customers = gender_counts.sum()
gender_percents = (gender_counts / total_customers) * 100

# Print the segment counts and percentages
print("Gender counts:\n", gender_counts)
print("\nGender percentages:\n", gender_percents.apply(lambda x: '{:.2f}%'.format(x)))

Gender counts:
 Female    515572
Male      434037
Name: gender, dtype: int64

Gender percentages:
 Female    54.29%
Male      45.71%
Name: gender, dtype: object


In [44]:
# Count the number of customers in each segment using value_counts()
segment_counts = df_sorted['customer_segment'].value_counts()

# Calculate percentage of customers in each segment
total_customers = segment_counts.sum()
segment_percents = (segment_counts / total_customers) * 100

# Print the segment counts and percentages
print("Segment counts:\n", segment_counts)
print("\nSegment percentages:\n", segment_percents.apply(lambda x: '{:.2f}%'.format(x)))

Segment counts:
 Individuals         551435
College Graduate    350480
VIP                  36009
Name: customer_segment, dtype: int64

Segment percentages:
 Individuals         58.79%
College Graduate    37.37%
VIP                  3.84%
Name: customer_segment, dtype: object


In [45]:
# Count the number of active and inactive customers
activity_counts = df_sorted['activity_index'].value_counts()

# Rename the index labels from 0.0 to 'Inactive' and 1.0 to 'Active'
activity_counts = activity_counts.rename({0.0: 'Inactive', 1.0: 'Active'})

# Calculate percentage of active and inactive customers
total_customers = activity_counts.sum()
activity_percents = (activity_counts / total_customers) * 100

# Print the activity counts and percentages
print("Activity counts:\n", activity_counts)
print("\nActivity percentages:\n", activity_percents.apply(lambda x: '{:.2f}%'.format(x)))


Activity counts:
 Inactive    552638
Active      396971
Name: activity_index, dtype: int64

Activity percentages:
 Inactive    58.20%
Active      41.80%
Name: activity_index, dtype: object


In [40]:
# Calculate the average age of customers
average_age = round(df_sorted['age'].mean())

# Print the average age
print("Average age of customers:", average_age)

Average age of customers: 40


In [42]:
# Creating age group and determining the number of customers in each age group
# Define age group bins and labels
bins = [0, 18, 25, 40, 60, np.inf]
labels = ['0-18', '19-25', '26-40', '41-60', '60+']

# Create a new column with the age group labels
df_sorted['age_group'] = pd.cut(df_sorted['age'], bins=bins, labels=labels)

# Count the number of customers in each age group using value_counts()
age_counts = df_sorted['age_group'].value_counts()

# Calculate percentage of customers in each age group
total_customers = age_counts.sum()
age_percents = (age_counts / total_customers) * 100

# Print the age counts and percentages
print("Age group counts:\n", age_counts)
print("\nAge group percentages:\n", age_percents.apply(lambda x: '{:.2f}%'.format(x)))

Age group counts:
 41-60    315840
19-25    264921
26-40    239993
60+      120651
0-18       8204
Name: age_group, dtype: int64

Age group percentages:
 41-60    33.26%
19-25    27.90%
26-40    25.27%
60+      12.71%
0-18      0.86%
Name: age_group, dtype: object


In [49]:
# Get the counts of customers by country
country_counts = df_sorted['country_of_residence'].value_counts()

# Get the top 10 countries and their counts
top_countries = country_counts[:10]

# Combine the remaining countries into an 'Others' category
other_count = country_counts[10:].sum()
top_countries['Others'] = other_count

# Calculate the percentage of customers in each country
total_customers = country_counts.sum()
country_percents = (top_countries / total_customers) * 100

# Print the top 10 countries and their counts and percentages
print("Top 10 countries of residence:\n", top_countries)
print("\nPercentage of customers in each country:\n", country_percents.apply(lambda x: '{:.2f}%'.format(x)))

Top 10 countries of residence:
 ES        945592
FR           317
AR           289
GB           286
DE           285
US           230
CO           211
IT           180
RO           175
MX           157
Others      1887
Name: country_of_residence, dtype: int64

Percentage of customers in each country:
 ES        99.58%
FR         0.03%
AR         0.03%
GB         0.03%
DE         0.03%
US         0.02%
CO         0.02%
IT         0.02%
RO         0.02%
MX         0.02%
Others     0.20%
Name: country_of_residence, dtype: object


99.58% of customers are resident in Spain

In [51]:
# Calculate the average income
average_income = df_sorted['gross_household_income'].mean()

# Print the average income
print("Average income:", average_income.round())


Average income: 134096.0


Average Gross Household Income = $134,096

In [61]:
# Count of customers by gross household income bands
# Define bins for gross_household_income
bins = [0, 50000, 100000, 150000, 200000, 300000, 500000, 1000000, 5000000, 10000000, 50000000]

# Create a new column with the income group labels
labels = [f"{bins[i]:,.0f} - {bins[i+1]:,.0f}" for i in range(len(bins)-1)]
df_sorted['income_group'] = pd.cut(df_sorted['gross_household_income'], bins=bins, labels=labels)

# Count the number of customers in each income group using value_counts()
income_counts = df_sorted['income_group'].value_counts()

# Calculate percentage of customers in each income group
total_customers = income_counts.sum()
income_percents = (income_counts / total_customers) * 100

# Print the income counts and percentages
print("Income group counts:\n", income_counts)
print("\nIncome group percentages:\n", income_percents.apply(lambda x: '{:.2f}%'.format(x)))


Income group counts:
 50,000 - 100,000           272061
100,000 - 150,000          171483
150,000 - 200,000           85829
0 - 50,000                  75838
200,000 - 300,000           65826
300,000 - 500,000           29149
500,000 - 1,000,000          7442
1,000,000 - 5,000,000        1579
5,000,000 - 10,000,000        127
10,000,000 - 50,000,000        74
Name: income_group, dtype: int64

Income group percentages:
 50,000 - 100,000           38.35%
100,000 - 150,000          24.17%
150,000 - 200,000          12.10%
0 - 50,000                 10.69%
200,000 - 300,000           9.28%
300,000 - 500,000           4.11%
500,000 - 1,000,000         1.05%
1,000,000 - 5,000,000       0.22%
5,000,000 - 10,000,000      0.02%
10,000,000 - 50,000,000     0.01%
Name: income_group, dtype: object


In [62]:
#Customer that joined within the last 6 months
new_customer_counts = df_sorted['new_customer'].value_counts()

# Rename the index labels from 0.0 to 'Old Customers' and 1.0 to 'New Customers'
new_customer_counts = new_customer_counts.rename({0.0: 'Old Customers', 1.0: 'New Customers'})

# Calculate percentage of active and inactive customers
total_customers = new_customer_counts.sum()
new_customer_percents = (new_customer_counts / total_customers) * 100

# Print the new customer counts and percentages
print("New customer counts:\n", new_customer_counts)
print("\nNew customer percentages:\n", new_customer_percents.apply(lambda x: '{:.2f}%'.format(x)))

New customer counts:
 Old Customers    906388
New Customers     43221
Name: new_customer, dtype: int64

New customer percentages:
 Old Customers    95.45%
New Customers     4.55%
Name: new_customer, dtype: object


4.55% of the customers joined within the last 6 months

In [73]:
# Count number of customers with a current account
num_current_acc = df_sorted['current_acc'].sum()

# Calculate percentage of customers with a current account
total_customers = len(df_sorted)
percent_current_acc = (num_current_acc / total_customers) * 100

# Print results
print("Number of customers with a current account:", num_current_acc)
print("Percentage of customers with a current account: {:.2f}%".format(percent_current_acc))


Number of customers with a current account: 561810
Percentage of customers with a current account: 59.16%


In [76]:
# Count of customers with each product or account type  and the percentage of total
# Define the list of account types
account_types = ['current_acc', 'saving_acc', 'derivative_acc', 'guarantees', 'payroll_acc', 'junior_acc', 'more_particular_acc', 
                 'particular_acc', 'particular_plus_acc', 'short_term_deposits', 'medium_term_deposits', 'long_term_deposits', 
                 'e_account', 'funds', 'mortgage', 'pensions_plan', 'loans', 'taxes', 'credit_card', 'securities', 'home_acc', 
                 'payroll', 'pensions', 'direct_debit']

# Initialize a dictionary to store the number of customers for each account type
account_counts = {}

# Loop over the account types and count the number of customers for each type using sum()
for account in account_types:
    account_counts[account] = df_sorted[account].sum()

# Calculate the percentage of customers with each account type
total_customers = len(df_sorted)
account_percents = {account: (count / total_customers) * 100 for account, count in account_counts.items()}

# Print the account counts and percentages
for account in account_types:
    print(f"{account.capitalize()} accounts:")
    print(f"Number of customers: {account_counts[account]}")
    print(f"Percentage of total: {account_percents[account]:.2f}%\n")


Current_acc accounts:
Number of customers: 561810
Percentage of total: 59.16%

Saving_acc accounts:
Number of customers: 78
Percentage of total: 0.01%

Derivative_acc accounts:
Number of customers: 316
Percentage of total: 0.03%

Guarantees accounts:
Number of customers: 16
Percentage of total: 0.00%

Payroll_acc accounts:
Number of customers: 73067
Percentage of total: 7.69%

Junior_acc accounts:
Number of customers: 7587
Percentage of total: 0.80%

More_particular_acc accounts:
Number of customers: 8097
Percentage of total: 0.85%

Particular_acc accounts:
Number of customers: 100603
Percentage of total: 10.59%

Particular_plus_acc accounts:
Number of customers: 33274
Percentage of total: 3.50%

Short_term_deposits accounts:
Number of customers: 318
Percentage of total: 0.03%

Medium_term_deposits accounts:
Number of customers: 1021
Percentage of total: 0.11%

Long_term_deposits accounts:
Number of customers: 31218
Percentage of total: 3.29%

E_account accounts:
Number of customers: 7

In [79]:
# Define the account columns
account_cols = ['current_acc', 'saving_acc', 'derivative_acc', 'guarantees', 'payroll_acc', 'junior_acc', 'more_particular_acc', 'particular_acc', 'particular_plus_acc', 'short_term_deposits', 'medium_term_deposits', 'long_term_deposits', 'e_account', 'funds', 'mortgage', 'pensions_plan', 'loans', 'taxes', 'credit_card', 'securities', 'home_acc', 'payroll', 'pensions', 'direct_debit']

# Count the number of customers with each account type
account_counts = df_sorted[account_cols].sum()

# Calculate percentages of total
total_customers = len(df_sorted)
account_percents = (account_counts / total_customers) * 100

# Combine counts and percentages into a single DataFrame
account_df = pd.DataFrame({'counts': account_counts, 'percentages': account_percents.apply(lambda x: '{:.2f}%'.format(x))})

# Sort the DataFrame by count and percentage in descending order
account_df = account_df.sort_values(['counts', 'percentages'], ascending=False)

# Print the results
print("Counts by account type:\n", account_df['counts'])
print("\nPercentages by account type:\n", account_df['percentages'])


Counts by account type:
 current_acc             561810.0
direct_debit            113463.0
particular_acc          100603.0
e_account                75866.0
payroll_acc              73067.0
pensions                 52909.0
payroll                  48361.0
taxes                    45890.0
credit_card              34826.0
particular_plus_acc      33274.0
long_term_deposits       31218.0
securities               21426.0
funds                    14688.0
more_particular_acc       8097.0
junior_acc                7587.0
pensions_plan             7359.0
mortgage                  4527.0
home_acc                  2960.0
loans                     1994.0
medium_term_deposits      1021.0
short_term_deposits        318.0
derivative_acc             316.0
saving_acc                  78.0
guarantees                  16.0
Name: counts, dtype: float64

Percentages by account type:
 current_acc             59.16%
direct_debit            11.95%
particular_acc          10.59%
e_account                7.99%

In [86]:
# Get the column index number of saving_acc (the first product in the columns)
saving_acc_index = df_sorted.columns.get_loc('saving_acc')

print(saving_acc_index)

24


In [97]:
# Count the number of banking products offered by the bank
bank_products = df_sorted.iloc[:, saving_acc_index:].shape[1]

print("Number of banking products:", bank_products)

Number of banking products: 27


In [88]:
# Count of banking products owned by customers.
num_products = df_sorted.iloc[:, saving_acc_index:].sum(axis=1)

# Count the number of customers with each number of bank products
num_customers = num_products.value_counts()

# Calculate the percentage of customers with each number of bank products
total_customers = num_customers.sum()
percent_customers = (num_customers / total_customers) * 100

# Print the results
print("Number of customers with each number of bank products:")
print(num_customers)
print("\nPercentage of customers with each number of bank products:")
print(percent_customers.apply(lambda x: '{:.2f}%'.format(x)))


Number of customers with each number of bank products:
1.0     453809
0.0     252104
2.0     120653
3.0      47830
4.0      28772
5.0      18046
6.0      12689
7.0       8356
8.0       4445
9.0       1897
10.0       692
11.0       239
12.0        62
13.0        12
14.0         2
15.0         1
dtype: int64

Percentage of customers with each number of bank products:
1.0     47.79%
0.0     26.55%
2.0     12.71%
3.0      5.04%
4.0      3.03%
5.0      1.90%
6.0      1.34%
7.0      0.88%
8.0      0.47%
9.0      0.20%
10.0     0.07%
11.0     0.03%
12.0     0.01%
13.0     0.00%
14.0     0.00%
15.0     0.00%
dtype: object


C:\Users\Oludipe j\AppData\Local\Temp\ipykernel_11756\1841823268.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  num_products = df_sorted.iloc[:, saving_acc_index:].sum(axis=1)


In [46]:
df_sorted.head(3)

data_date  customer_id employee_index country_of_residence gender  \
13026343  2016-05-28        15889              F                   ES      F   
13026342  2016-05-28        15890              A                   ES      F   
5319232   2015-08-28        15891              N                   ES      M   

           age date_joined  new_customer customer_seniority  primary_index  \
13026343  56.0  1995-01-16           0.0                255            1.0   
13026342  63.0  1995-01-16           0.0                256            1.0   
5319232   59.0  2015-07-28           0.0                246           99.0   

         last_date_as_primary customer_type customer_relation resident  \
13026343                  NaN             1                 A      Yes   
13026342                  NaN             1                 A      Yes   
5319232            2015-08-05             1                 A      Yes   

         foreigner employee_spouse channel deceased  address_type  \
13026343        No               N     KAT       No           1.0   
13026342        No               N     KAT       No           1.0   
5319232         No               N     KAT       No           1.0   

          province_code province_name  activity_index  gross_household_income  \
13026343           28.0        MADRID             1.0                326124.9   
13026342           28.0        MADRID             1.0                 71461.2   
5319232            28.0        MADRID             0.0                     NaN   

         customer_segment  saving_acc  guarantees  current_acc  \
13026343              VIP           0           0            1   
13026342              VIP           0           0            0   
5319232       Individuals           0           0            0   

          derivative_acc  payroll_acc  junior_acc  more_particular_acc  \
13026343               0            0           0                    0   
13026342               0            1           0                    0   
5319232                0            0           0                    0   

          particular_acc  particular_plus_acc  short_term_deposits  \
13026343               0                    1                    0   
13026342               0                    1                    0   
5319232                0                    0                    0   

          medium_term_deposits  long_term_deposits  e_account  funds  \
13026343                     0                   0          0      0   
13026342                     0                   0          1      0   
5319232                      0                   0          0      0   

          mortgage  pensions_plan  loans  taxes  credit_card  securities  \
13026343         0              0      0      0            1           1   
13026342         0              1      0      0            1           0   
5319232          0              0      0      0            0           0   

          home_acc  payroll  pensions  direct_debit data_datetime age_group  
13026343         0      0.0       0.0             0    2016-05-28     41-60  
13026342         0      1.0       1.0             1    2016-05-28       60+  
5319232          0      0.0       0.0             0    2015-08-28     41-60

## **Recommendations:**

Based on the analysis, here are some recommendations to increase cross-selling and revenue at XYZ Credit Union:
1.	Analyze customer behavior: Conduct a thorough analysis of customer behavior to understand their preferences and identify patterns that can be used to drive cross-selling. This could include analyzing customer demographics, transaction history, and product ownership data to determine which products are most popular among which customer segments.
2.	Develop targeted marketing campaigns: Based on the insights gained from customer behavior analysis, develop targeted marketing campaigns to promote cross-selling of specific products to relevant customer segments. This could include personalized offers, incentives, and rewards to encourage customers to purchase additional products.
3.	Train employees on cross-selling: Ensure that employees are trained in cross-selling strategies and are equipped with the knowledge and skills to effectively promote additional products to customers. This could include offering incentives for successful cross-selling and providing ongoing training and support to employees.
4.	Streamline product offerings: Consider streamlining the bank's product offerings to reduce complexity and make it easier for customers to understand and purchase multiple products. This could include bundling products together or simplifying product names and descriptions.
5.	Improve customer experience: Focus on improving the overall customer experience to build trust and increase loyalty, which can lead to increased cross-selling opportunities. This could include providing personalized service, offering easy-to-use online and mobile banking tools, and responding quickly to customer inquiries and issues.

By implementing these recommendations, XYZ Credit Union can improve cross-selling and increase revenue from existing customers.

